### ISSUE TITLE
IntTaxonomyFacets chooses dense values array when FacetsCollector has no MatchingDocs

### ISSUE THREAD

**Issue by Shradha26** [apache/lucene#12558]
### Description

I recently encountered an Out of Memory bug when `IntTaxonomyFacets` initialised a dense values array instead of a sparse hash map. This was happening because we encountered a case where `sumTotalHits` and `maxDoc` were both 0, and since 0<0 is false, `useHashTable()` i.e. the method we call to decide if we should use a sparse/dense data structure, returns false and we end up using the dense values array. 

```
 // if our result set is < 10% of the index, we collect sparsely (use hash map):
sumTotalHits < maxDoc / 10;
```

We arrived at the `maxDoc = 0` case when a `FacetsCollector` that was not null and had `totalHits = 0` but an empty `List<MatchingDocs>` was passed to the method. [We end up not iterating the loop and consequently not updating maxDoc value](https://github.com/apache/lucene/blob/branch_9_7/lucene/facet/src/java/org/apache/lucene/facet/taxonomy/IntTaxonomyFacets.java#L120) and it stays 0. 

I think we could explicitly check for `maxDoc == 0` in the condition too, but maybe there's a better fix?

**Comment by gsmiller** [comment]
Thanks for opening this issue @Shradha26! Do we have a test case that reproduces this? I'm still a little confused on how we can actually arrive in this state?

**Comment by Shradha26** [comment]
Hey @gsmiller, thanks for looking at the issue. Here's a draft PR for a test case that replicates this state: https://github.com/apache/lucene/pull/12563. The test ``testMaxDocIsNeverZero`` does not pass because I throw an ``IllegalStateException`` from ``IntTaxonomyFacets`` when ``maxDoc==0``. I'm seeing this happen when I use a ``FacetsCollector`` that was returned in a ``DrillSideways.search`` result for counting.
I'm able to replicate this behaviour in ``IntTaxonomyFacets`` in the `branch_9_7` since it does not lazily initialise facet counting data structures. (The ``IntTaxonomyFacets`` on ``main`` branch does.)

**Comment by gsmiller** [comment]
Oh interesting. Thanks @Shradha26! Is drill-sideways the only means for reproducing this (that we're aware of anyway)? That's tricky. Do we think it's doing the right thing, or should we be changing drill-sideways to ensure this situation never arises?

**Comment by gsmiller** [comment]
OK, came back across this while cleaning up open browser tabs and decided to repro it myself. I know what's going on. It has to do with `#finish` not properly getting called on sideways facet collectors when the base query matches no docs and short-circuits. I've verified a fix locally, but it's a bit crude. I'll think about the best way to solve this a little more and put up a PR.

**Comment by gsmiller** [comment]
I opened a PR here that fixes the root cause of this bug: https://github.com/apache/lucene/pull/12853

**Comment by gsmiller** [comment]
Fixed the root cause of this in #12853
